iox_catalog/src/postgres.rs

```diff
@@ -0,0 +1,874 @@
+        new_columns: BTreeMap<i32, BTreeMap<String, ColumnSchema>>,
+    ) {
+        for (table_name, table_id) in new_tables {
+            self.tables
+                .entry(table_name)
+                .or_insert_with(|| TableSchema::new(table_id));
+        }
+
+        for (table_id, new_columns) in new_columns {
+            let table = self
+                .get_table_mut(table_id)
+                .expect("table must be in namespace to add columns");
+            table.add_columns(new_columns);
+        }
+    }
+
+    fn get_table_mut(&mut self, table_id: i32) -> Option<&mut TableSchema> {
+        for (_, table) in self.tables.iter_mut() {
+            if table.id == table_id {
+                return Some(table);
+            }
+        }
+
+        None
+    }
+}
+
+/// Column definitions for a table
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct TableSchema {
+    /// the table id
+    pub id: i32,
+    /// the table's columns by their name
+    pub columns: BTreeMap<String, ColumnSchema>,
+}
+
+impl TableSchema {
+    pub fn new(id: i32) -> Self {
+        Self {
+            id,
+            columns: BTreeMap::new(),
+        }
+    }
+
+    fn add_columns(&mut self, columns: BTreeMap<String, ColumnSchema>) {
+        for (name, column) in columns {
+            self.columns.insert(name, column);
+        }
+    }
+}
+
+/// The column id and its type for a column
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ColumnSchema {
+    /// the column id
+    pub id: i32,
+    /// the column type
+    pub data_type: ColumnType,
+}
+
+impl ColumnSchema {
+    fn is_tag(&self) -> bool {
+        self.data_type == ColumnType::Tag
+    }
+
+    fn matches_field_type(&self, field_value: &FieldValue) -> bool {
+        match (field_value, self.data_type) {
+            (FieldValue::I64(_), ColumnType::I64) => true,
+            (FieldValue::U64(_), ColumnType::U64) => true,
+            (FieldValue::F64(_), ColumnType::F64) => true,
+            (FieldValue::String(_), ColumnType::String) => true,
+            (FieldValue::Boolean(_), ColumnType::Bool) => true,
+            _ => false,
+        }
+    }
+}
+
+/// The column data type
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum ColumnType {
+    I64 = 1,
+    U64 = 2,
+    F64 = 3,
+    Bool = 4,
+    String = 5,
+    Time = 6,
+    Tag = 7,
+}
+
+impl ColumnType {
+    fn as_str(&self) -> &str {
+        match self {
+            ColumnType::I64 => "i64",
+            ColumnType::U64 => "u64",
+            ColumnType::F64 => "f64",
+            ColumnType::Bool => "bool",
+            ColumnType::String => "string",
+            ColumnType::Time => "time",
+            ColumnType::Tag => "tag",
+        }
+    }
+}
+
+impl std::fmt::Display for ColumnType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let s = self.as_str();
+
+        write!(f, "{}", s)
+    }
+}
+
+impl TryFrom<i16> for ColumnType {
+    type Error = ();
```

Comment:
```suggestion
    type Error = Box<dyn std::error::Error>;
```

and then swap the `Err(())` tbelow o be:

```rust
_ => Err("invalid column value".into()),
```

This way callers will still have a useful error message to display/log and you don't need to make another type.